File: app/modules/zen-remote-desktop/android/src/main/java/expo/modules/zenremotedesktop/DesktopTransportPolicy.kt
package expo.modules.zenremotedesktop

import java.net.URI
import okhttp3.HttpUrl
import okhttp3.Request

internal object DesktopTransportPolicy {
  fun privateAddress(bytes: ByteArray): Boolean {
    val b = bytes.map { it.toInt() and 255 }
    if (b.size == 16 && b.take(10).all { it == 0 } && b[10] == 255 && b[11] == 255) {
      return privateAddress(bytes.copyOfRange(12, 16))
    }
    return if (b.size == 4) b[0] == 10 || (b[0] == 172 && b[1] in 16..31) ||
      (b[0] == 192 && b[1] == 168) || (b[0] == 100 && b[1] in 64..127)
    else b.size == 16 && (b[0] and 254) == 252
  }

  private fun parse(value: String, endpoint: Boolean): HttpUrl? = try {
    val uri = URI(value)
    if (uri.scheme !in listOf("ws", "wss") || uri.rawUserInfo != null || uri.rawQuery != null || uri.rawFragment != null ||
        (if (endpoint) uri.rawPath != "/desktop" else uri.rawPath !in listOf("", "/"))) null
    else Request.Builder().url(value).build().url
  } catch (_: Exception) { null }

  private fun sameOrigin(a: HttpUrl, b: HttpUrl) = a.scheme == b.scheme && a.host == b.host && a.port == b.port

  fun allows(url: String, mode: String, boundOrigin: String, sourceOrigin: String, pin: String,
             numericAddress: (String) -> ByteArray?): Boolean {
    val target = parse(url, true) ?: return false
    val bound = parse(boundOrigin, false) ?: return false
    val source = parse(sourceOrigin, false) ?: return false
    if (!sameOrigin(target, bound)) return false
    return when (mode) {
      "tls" -> target.isHttps && source.isHttps && sameOrigin(target, source)
      "pinned-link" -> !target.isHttps && target.host == "127.0.0.1" && source.isHttps &&
        pin.length == 64 && pin.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }
      "trusted-lan" -> !target.isHttps && sameOrigin(target, source) &&
        !target.host.contains('%') && numericAddress(target.host)?.let(::privateAddress) == true
      else -> false
    }
  }
}

File: app/modules/zen-remote-desktop/ios/DesktopTransportPolicy.swift
import Foundation
import Network

enum DesktopTransportPolicy {
  private static func privateAddress(_ bytes: [UInt8]) -> Bool {
    if bytes.count == 16 && bytes.prefix(10).allSatisfy({ $0 == 0 }) && bytes[10] == 255 && bytes[11] == 255 {
      return privateAddress(Array(bytes.suffix(4)))
    }
    if bytes.count == 4 {
      return bytes[0] == 10 || (bytes[0] == 172 && (16...31).contains(bytes[1])) ||
        (bytes[0] == 192 && bytes[1] == 168) || (bytes[0] == 100 && (64...127).contains(bytes[1]))
    }
    return bytes.count == 16 && (bytes[0] & 254) == 252
  }

  private static func parse(_ value: String, endpoint: Bool) -> URLComponents? {
    guard let parsed = URLComponents(string: value), ["ws", "wss"].contains(parsed.scheme ?? ""),
      parsed.host != nil, parsed.user == nil, parsed.password == nil,
      parsed.query == nil, parsed.fragment == nil,
      (endpoint ? parsed.percentEncodedPath == "/desktop" : ["", "/"].contains(parsed.percentEncodedPath)),
      parsed.port == nil || (1...65535).contains(parsed.port!) else { return nil }
    return parsed
  }

  private static func sameOrigin(_ a: URLComponents, _ b: URLComponents) -> Bool {
    a.scheme == b.scheme && a.host?.lowercased() == b.host?.lowercased() &&
      (a.port ?? (a.scheme == "wss" ? 443 : 80)) == (b.port ?? (b.scheme == "wss" ? 443 : 80))
  }

  static func allows(_ config: [String: String]) -> Bool {
    guard let target = parse(config["url"] ?? "", endpoint: true),
      let bound = parse(config["boundOrigin"] ?? "", endpoint: false),
      let source = parse(config["sourceOrigin"] ?? "", endpoint: false), sameOrigin(target, bound) else { return false }
    switch config["transport"] {
    case "tls": return target.scheme == "wss" && sameOrigin(target, source)
    case "pinned-link":
      let pin = config["transportPin"] ?? ""
      return target.scheme == "ws" && target.host == "127.0.0.1" && source.scheme == "wss" &&
        pin.utf8.count == 64 && pin.utf8.allSatisfy { (48...57).contains($0) || (65...70).contains($0) || (97...102).contains($0) }
    case "trusted-lan":
      guard target.scheme == "ws", sameOrigin(target, source), let rawHost = target.host, !rawHost.contains("%") else { return false }
      let host = rawHost.hasPrefix("[") && rawHost.hasSuffix("]") ? String(rawHost.dropFirst().dropLast()) : rawHost
      if let address = IPv4Address(host) { return privateAddress(Array(address.rawValue)) }
      if let address = IPv6Address(host) { return privateAddress(Array(address.rawValue)) }
      return false
    default: return false
    }
  }
}

final class DesktopNoRedirectDelegate: NSObject, URLSessionTaskDelegate {
  func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
                  newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
    completionHandler(nil)
  }
}
